pkg/pillar/cmd/domainmgr/domainmgr.go

```diff
@@ -1116,6 +1149,152 @@ func lookupDomainConfig(ctx *domainContext, key string) *types.DomainConfig {
 	return &config
 }
 
+func getHostCpusNum() (int, error) {
+	// The `nproc` util relies on the /proc/sys file system, that is not properly accessible from a container.
+	// So, have to run the command with access to the host FS.
+	out, err := exec.Command("chroot", "/hostfs", "nproc", "--all").Output()
+	if err != nil {
+		return -1, err
+	}
+	// remove "\n"
+	out = out[:2]
+	maxHostCpus, _ := strconv.Atoi(string(out))
+	return maxHostCpus, nil
+}
+
+func setCgroupCpuset(vmUUIDAndVersion string, cpus string) error {
+	cpuSetFileName := fmt.Sprintf("/sys/fs/cgroup/cpuset/eve-user-apps/%s/cpuset.cpus", vmUUIDAndVersion)
+	subCommand := fmt.Sprintf("echo %s > %s", cpus, cpuSetFileName)
+	// Has to run with access to host FS. The corresponding cpuset directory is not accessible from within a container
+	err := exec.Command("chroot", "/hostfs", "sh", "-c", subCommand).Run()
+	if err != nil {
+		log.Warnf("Failed to execute: 'chroot /hostfs sh -c %s'", subCommand)
+		return err
+	}
+	return nil
+}
+
+func constructNonPinnedCpumaskString(ctx *domainContext) string {
+	result := ""
+	for cpu, pinned := range ctx.hostCpusPinned {
+		if !pinned {
+			addToMask(cpu, &result)
+		}
+	}
+	return result
+}
+
+func addToMask(cpu int, s *string) {
+	if s == nil {
+		return
+	}
+	if *s == "" {
+		*s = fmt.Sprintf("%d", cpu)
+	} else {
+		*s = fmt.Sprintf("%s,%d", *s, cpu)
+	}
+}
+
+func excludeCPUFromNonPinnedVMs(ctx *domainContext, cpu int) error {
+	log.Functionf("Pinning a CPU (%d), exclude it from the other VMs CPU sets...", cpu)
+	ctx.hostCpusPinned[cpu] = true
+	//Exclude the CPU from the shared list and fix the cgroups of previously created VMs
+	for vmUUID := range ctx.resourcesUsedByVMs {
+		vmToCheck := ctx.resourcesUsedByVMs[vmUUID]
+		// Need the Status here only to get the proper Key to find the corresponding Config
+		vmStatus := lookupDomainStatusByUUID(ctx, vmUUID)
+		if vmStatus == nil {
+			//Most probably the status has not been created yet
+			continue
+		}
+		if vmStatus.VmConfig.CPUsPinned {
+			// We need to check only non-pinned VMs
+			continue
+		}
+		vmConfig := lookupDomainConfig(ctx, vmStatus.Key())
+		if vmConfig == nil {
+			// Strange situation. Should be investigated if happens.
+			log.Warnf("Failed to find the DomainConfig corresponding to a DomainStatus (%s)", vmStatus.DisplayName)
+			continue
+		}
+		cpuSetChanged := false
+		for cpuToCheck, used := range vmToCheck.CPUSet {
+			if cpuToCheck == cpu && used {
+				// Found previously assigned CPU. Have to exclude it
+				vmToCheck.CPUSet[cpu] = false
+				cpuSetChanged = true
+			}
+		}
+		// If at least one change in the CPU set of the VM happened, adapt the corresponding cgroup
+		if cpuSetChanged {
+			vmConfig.VmConfig.CPUs = constructNonPinnedCpumaskString(ctx)
+			err := setCgroupCpuset(vmConfig.GetTaskName(), vmConfig.VmConfig.CPUs)
+			if err != nil {
+				return fmt.Errorf("failed to redistribute CPUs between VMs, can affect the inter-VM isolation")
+			}
+		}
```

Comment:
It looks like we should have a log message for this change.

Reply:
Will do

Reply:
done